### DISCUSSION TITLE
Outputs

### DISCUSSION THREAD

**Discussion by baritonomarchetto** [djyt/cannonball#83] [Q&A]
Hi Chris, by looking at the code I noticed that not only motor outputs are implemented, but also lamps. In mame we can use mamehooker to send lamps states to external hardware: how can this be done in cannonball?
(on a side note : I am just willing a check point lamp in outrun like the one in outrunners, hehe...)

**Comment by djyt** [comment]
Yes, they are all hooked up and work on a real cabinet using the SmartyPi interface which is currently at prototype phase.

The outputs are currently written to the console everytime they change. Try enabling the smartypi section in the configuration and you'll see what I mean. On the SmartyPi hardware itself, an Arduino sits there reading the console and outputs to hardware. 

At one point we had CannonBall working with a full sitdown cab, so even the full deluxe code is ported and working.

I've never used Mamehooker - so I don't know if it is compatible with the way I've done things. Feel free to investigate and let me know. 

I'm currently working on supporting all the chip-based music from the Switch and 3DS versions. Works great, the main problem is actually extracting the data from the binaries is a relatively complex process for a typical person to do themselves.

**Reply by baritonomarchetto** [reply]
I have opened a post on BYOAC just to see what the mamehooker creator think of this and ask him if it makes more sense to implement some code on cannonball or mamehooker itself.
I am an arduino user and it would be cool to see the code you used to read the console
(Thanks!)

**Reply by djyt** [reply]
With SmartyPi, both CannonBall and MAME are run from a python wrapper. This wrapper essentially sits in a loop capturing the output from the console and sending it to the Arduino. The Arduino, in turn, writes to the appropriate hardware. Both MAME and CannonBall have been modified to use identical outputs so they can be interchanged. 

Not sure if the source code will be released for SmartyPi as it's not my project. But hopefully.

**Reply by baritonomarchetto** [reply]
So, as far as you know, you are using a modified mame in order to have the wrapper read the consolle, or it is a standard mame? Because in the second case being that mamehooker reads without problems outputs from mame, I think it could be an "easy" task to read also cannonball outs.

**Reply by djyt** [reply]
It's a modified version of MAME to output various information to the console. Although literally a few lines per driver. Bear in mind SmartyPi is tailored to support a small number of machines.

The python wrapper reads the console and sends it to the Arduino.

What are you looking to do with Mame hooker and the outputs? Some sort of funky light show? :)

**Reply by baritonomarchetto** [reply]
Hehe, no no: it's only that sega games without the flashing start button are not the same :)
Also, in case you add some new light "show" (check point flashing? Who knows :)) I want to be ready

**Reply by djyt** [reply]
A modified MAME. 

For MameHooker - the easy thing to do is just take the cannonball source code and modify it to the format it expects (disclaimer: I have never looked at MameHooker!)

I'd start by looking at the outputs to the console here:
https://github.com/djyt/cannonball/blob/master/src/main/engine/ooutputs.cpp
